cs3/storage/provider/v1beta1/resources.proto

```diff
@@ -189,27 +190,29 @@ message ResourceChecksumPriority {
   uint32 priority = 2;
 }
 
-// The mechanism to identify a resource
-// in the storage provider namespace. Note that the path OR the resourceId must be specifed, not both.
+// The mechanism to identify a resource in the CS3 namespace.
+// It can represent path based, id based and combined references:
+// The storage registry uses the storage_id to determine the responsible storage provider.
+// When the storage_id is not available it will use the path.
+// In a URL the different components can be represented in a string using the following layout:
+// <storage_id>!<node_id>:<path>
 message Reference {
-  // REQUIRED.
-  // One of the specifications MUST be specified.
-  oneof spec {
-    // The path to the resource.
-    // MUST start with the slash character (/).
-    string path = 1;
-    // The id for the resource.
-    // MUST NOT start with the slash character (/).
-    ResourceId id = 2;
-  }
+  // OPTIONAL
+  ResourceId resource_id = 1;
```

Comment:
There is one point I don't understand here. From an API perspective this is a bit inconsistent. If both properties are optional, then an empty Reference should be valid, by definition, but an empty reference is not helpful at all. In the previous design this was ensured not to happen because it was defined as a `oneof` relationship:

> If you have a message with many fields and where at most one field will be set at the same time, you can enforce this behavior and save memory by using the oneof feature.

I would update the docs to reflect that one MUST be present.

Reply:
+1